src/Razor/src/Microsoft.AspNetCore.Razor.VSCode.BlazorWasmDebuggingExtension/scripts/retrieveDebugProxy.js

```diff
@@ -9,66 +9,82 @@ const extract = require('extract-zip');
 const os = require('os');
 
-var finished = util.promisify(stream.finished);
+const finished = util.promisify(stream.finished);
 
 const formatLog = (text) => `[${new Date()}] ${text}`;
 const log = (text) => console.log(formatLog(text));
 const logError = (text) => console.error(formatLog(text));
 
 async function downloadProxyPackage(version) {
-    var nugetUrl = 'https://api.nuget.org/v3-flatcontainer';
-    var packageName = 'Microsoft.AspNetCore.Components.WebAssembly.DevServer';
-
     const tmpDirectory = path.join(os.tmpdir(), 'blazorwasm-companion-tmp');
     if (!fs.existsSync(tmpDirectory)){
         fs.mkdirSync(tmpDirectory);
     }
-    const extractTarget = path.join(tmpDirectory, `extracted-${packageName}.${version}`);
 
-    const versionedPackageName = `${packageName.toLowerCase()}.${version}.nupkg`;
     // nuget.org requires the package name be lower-case
-    const downloadUrl = `${nugetUrl}/${packageName.toLowerCase()}/${version}/${versionedPackageName}`;
-    const downloadPath = path.join(tmpDirectory, versionedPackageName);
+    const nugetUrl = 'https://api.nuget.org/v3-flatcontainer';
+    const packageName = 'Microsoft.AspNetCore.Components.WebAssembly.DevServer'.toLowerCase();
+    const versionedPackageName = `${packageName}.${version}.nupkg`;
+    const downloadUrl = `${nugetUrl}/${packageName}/${version}/${versionedPackageName}`;
 
     // Download and save nupkg to disk
     log(`Fetching package from ${downloadUrl}...`);
     const response = await fetch(downloadUrl);
 
     if (!response.ok) {
         logError(`Failed to download ${downloadUrl}`);
-        return null;
+        throw new Error(`Unable to download BlazorDebugProxy: ${response.status} ${response.statusText}`);
     }
+
+    const downloadPath = path.join(tmpDirectory, versionedPackageName);
     const outputStream = fs.createWriteStream(downloadPath);
     response.body.pipe(outputStream);
+    await finished(outputStream);
 
     // Extract nupkg to extraction directory
     log(`Extracting NuGet package with directory...`)
-    await finished(outputStream);
+    const extractTarget = path.join(tmpDirectory, `extracted-${packageName}.${version}`);
     await extract(downloadPath, { dir: extractTarget });
     return extractTarget;
 }
 
 async function copyDebugProxyAssets(version) {
-    var extracted = await downloadProxyPackage(version);
-    if (!extracted) {
-        return;
+    const targetDirectory = path.join(__dirname, '..', 'BlazorDebugProxy');
+    const versionMarkerFile = path.join(targetDirectory, '_version');
+    if (fs.existsSync(targetDirectory) && fs.existsSync(versionMarkerFile)) {
+        const cachedVersion = fs.readFileSync(versionMarkerFile, { encoding: 'utf-8' });
+        if (cachedVersion === version) {
+            log(`Found up-to-date BlazorDebugProxy ${version}, nothing to do.`);
+            return;
+        }
+
+        log(`Cached BlazorDebugProxy ${cachedVersion} is not ${version}, downloading...`);
```

Comment:
Proposed alternative for this: can we store the DebugProxy artifacts into the following folder structure: `tools/BlazorDebugProxy/${version}` then check for folder presence/contents instead of reading/writing from the file?

Also has the benfit of being able to maintain multiple cached debug proxy versions at once which has the benefit of making it easier to validate E2Es and new versions of the debug proxy.

Reply:
Ooh, I like this :)

Reply:
Added! Notes:
* We lose the strong guarantee that the proxy extracted correctly. This _should_ be fine for CI purposes because the script will error out, but it can make it tricky to debug local development because the script will see the subdirectory and return early even though there was an extraction error the last time it was run. If desired, I can add back creating a marker file after everything finishes successfully that the script can check against.
* Added some supporting logic to the extension itself to look in the right subdirectory. Didn't bother to add error handling if debugProxyVersion doesn't exist because the build script will fail without it.